### ISSUE TITLE
[research] Investigate symlinks and hard links

### ISSUE THREAD

**Issue by jgsogo** [conan-io/conan#7093]
There are many issues that talk about managing file links in Conan. Some (all?) of them are gathered into this project https://github.com/conan-io/conan/projects/30.

We need to define how it will look like in Conan v2.0 and see what can we do without breaking in Conan v1.xx, but first of all we need to have a clear idea about links in different systems.

 * What kind of links are supported in Windows/Linux/Mac? Do they have the same meaning? Do they behave the same?
 * What do python zip/tar libraries do with these links? Are all of them supported in the specification or support depends on the implementation?
 * What happens when the compressed file is created in one OS and uncompressed in a different one? Are symlinks preserved?
 * Is python able to create/identify those different types of links on every OS? Do they return the same values for the same functions? Do python versions behave the same?

Conan related question:
 * Do we want to store links in the packages? Only links that point inside the package? 
 * Should Conan replace absolute paths by relative ones?
 * What checksum to store in the `manifest` file?

---

Please, add any other question that can be interesting. After answering these questions we can think about Conan v2.0 and see how Conan 1.xxx behaves differently, it should help us to progress on related issues.

**Comment by TheQwertiest** [comment]
Windows support for symlinks is a bit meh: it requires either permission elevation or an enabled developer mode (which is accessible only on Windows 10). It's also questionable if symlinks are supported anywhere other than NTFS.

Source: https://blogs.windows.com/windowsdeveloper/2016/12/02/symlinks-windows-10/

**Comment by jgsogo** [comment]
Conan 2.0 could _just package_ the files in the `package_folder` and provide tools for the user to sanitize them:
 * `tools.symlinks_sanitize(conanfile, raise_if_error=True)`: this tool could transform absolute symlinks into relative ones (if they point to files inside the package), and print an error (raise if configured) for links outside the package (but don't modify them).

**Comment by memsharded** [comment]
Should we consider this closed with the introduction of #7178?

There is this PR open https://github.com/conan-io/conan/pull/6827 that contain tests for this hard link issue. Should we integrate it and see if the new tool can provide a fix for it?

**Comment by jgsogo** [comment]
I think we need to answer questions from the first comment first. I want to know how Python deals with these links in different OSs and what zip/tar libraries do to them. IMO it would be a useful reference, maybe something to add to FAQ although it is not related to Conan codebase.

Then, we can decide that Conan does nothing by default (just zip whatever is in the package folder) and you need to use provided tools (#7178) if you need help with links. And this is something we need to write in the docs before closing related issues.

---

About #6827, I really don't know how the tool `fix_symlinks` behaves with hard links 😫

**Comment by TheQwertiest** [comment]
> I really don't know how the tool fix_symlinks behaves with hard links

Hard links are just normal files in 99% of relevant cases. So I don't think any special casing is needed.

**Comment by SSE4** [comment]
I am linking #7760 here.
I believe during the work on PR, I've researched the topic and will be able to answer all the questions.

> What kind of links are supported in Windows/Linux/Mac?

in general, there are two major kind of links supported in operation systems are symbolic links and hard links. Windows also supports so-called [junctions](https://docs.microsoft.com/en-us/windows/win32/fileio/hard-links-and-junctions), but as they are Windows-specific, I don't consider they worth implementing in conan.
major thing is that file system support is needed for links - e.g. you can created symbolic and hard links on EXT4 and NTFS, but not on FAT/FAT32 (in any operation system - it's FS limitation). therefore, implementation should be careful when it tries to create symbolic and hard links (e.g. during the unzipping) - creating links on network or USB flash storage will likely to fail, and will require some fallback or post-processing (e.g. replacing links with copies, or just removing them).
another important note - while symbolic links are supported for directories and files, hard links are usually supported for files only (otherwise directory hard links break directed acyclic graph and make a lot troubles for tools traversing directory trees).

> Do they have the same meaning?

concept is the same everywhere, symbolic links is just fail storing another location. hard-links are just multiple names for the same file (inode).

> Do they behave the same?

not exactly, there are implementation details. for instance, on Linux you cannot change permissions on symlink itself (`chmod` call will always apply on target), while on Mac you can change permissions for link itself, as well as for linked target.

> What do python zip/tar libraries do with these links? 

python's `zipfile` module doesn't handle symlinks at all, it has to be done manually, and methods like `extractall` don't process them correctly. hardlinks seems to be unsupported by the format itself.
on the other hand, python's `tarfile` module works gracefully with symbolic links and hard links (even on Windows!), preserving them.
in general, storing links requires creating a special entry in an archive assigning correct link attributes/flags, so compressing code has to be a bit more complex usually, but still it's not very hard to correctly save all the information to store these links.

> Are all of them supported in the specification or support depends on the implementation?

for `tar`, there is a [specification](https://www.gnu.org/software/tar/manual/html_node/Standard.html) which includes symbolic links and hard links, and usually all software implementations handle links gracefully.
for `zip`, situation is much worse, and particular implementations, such as Windows Explorer, WinZIP, may fail to correctly process symbolic links. also, it seems like hard-links are not supported by `zip` at all. [specification](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT) doesn't include links, but many implementations support them (e.g. https://trac.edgewall.org/attachment/ticket/8919/ZipDownload.patch).
as a general rule, I'd recommend to always prefer `tar` format, if possible, as it correctly stores permissions, links, timestamps, and any other file attributes, while `zip` has many limitations and implementation differences.

> What happens when the compressed file is created in one OS and uncompressed in a different one? Are symlinks preserved?

`tarfile`: if destination OS and file system supports symlinks, symlinks are preserved. if not, they are replaced by copies by default. if symlink target cannot be resolved during copying, exception will be thrown.
```
    def makelink(self, tarinfo, targetpath):
        """Make a (symbolic) link called targetpath. If it cannot be created
          (platform limitation), we try to make a copy of the referenced file
          instead of a link.
        """
```
`zipfile`: default implementation doesn't support links and ignores them, storing copies (if possible).

> Is python able to create/identify those different types of links on every OS? Do they return the same values for the same functions? Do python versions behave the same?

there are bugs in python implementations, I believe, mostly on Windows. methods like `os.path.islink`, `os.path.realpath` or `os.readlink` may fail or return incorrect result for some python versions. most of bugs are already fixed by Python 3.8. to be portable, we may need to use native API via `ctypes` module instead of relying on the `os` module implementation. if needed, I may link python bugs here.

> Do we want to store links in the packages? 

yes, we do, as they are vital for many scenarios. they not only save disk space significantly, but in many cases are required (e.g. .so files symbolic links provide compatibility layer for linkers).

> Only links that point inside the package?

no, links pointing outside (e.g. to `/dev/null` or similar) are also valid cases. in general - we need to store symlinks as is, even broken - e.g. cyclic, pointing outside, pointing to non-existing files/directories, etc. we may provide tools to "fix" symlinks and packager may use them, if needed, as post- or pre- process: convert relative symlinks to absolute or vice versa, replace symlinks with copies, remove symlinks altogether, or whatever else user defines as "fixing".

> Should Conan replace absolute paths by relative ones?

not by default. conan may provide two helpers, e.g. to convert relative symbolic links to absolute, and vice versa, to convert absolute symbolic links to relative. user may choose to use these helpers before packing or after unpacking, and user should be able to choose to which symlinks to apply. of course, user should be able to keep absolute and relative symlinks as is (default case), even if there is a risk to get invalid symbolic links after unpacking.
to summarize, I suppose conan by default should store everything "as is" (even broken things).

> What checksum to store in the manifest file?

I am not so familiar with implementation details, if checksum only includes file contents (e.g. not attributes like permissions, timestamps and so on), for symbolic links it makes sense to compute checksum for the link itself. e.g. if we have symlink to the "/dev/null", we obviously don't compute checksum for "/dev/null" file itself (which is impossible as it's infinite), but we instead compute checksum for string "/dev/null" (9 characters only).

> `tools.symlinks_sanitize`

I'd say sanitizer should only detect errors/warnings and report (e.g. print) them, or return list of symlinks and problems with them (e.g. cyclic, non-existing, etc). it shouldn't try to "fix" them somehow (at least I wouldn't expect that from the method name).

another tool may "fix" symlink, but I'd suggest more precise helpers and names, e.g. the following could be done for "fixing":
- replace symlinks with copies (if possible)
- remove all symlinks
- remove "invalid" symlinks (pointing to the non existing file? cyclic? something else?)
- convert absolute paths to relative
- convert relative paths to absolute

> Hard links are just normal files in 99% of relevant cases. So I don't think any special casing is needed.

you need to detect them, e.g. compare inodes before packaging. then, you'll need to make special records (attributes/flags) in package archive in order to restore hard-links correctly (e.g. avoid unnecessary copies).

**Comment by memsharded** [comment]
Solved in https://github.com/conan-io/conan/pull/13137 for 2.0 (beta.10 today)
